include/ppx/knob.h

```diff
@@ -0,0 +1,284 @@
+class Knob
+{
+public:
+    Knob(const std::string& displayName, const std::string& flagName, const std::string& flagDesc)
+        : mDisplayName(displayName), mFlagName(flagName), mFlagDesc(flagDesc) {}
+
+    std::string GetDisplayName() const { return mDisplayName; }
+    std::string GetFlagName() const { return mFlagName; }
+    std::string GetFlagDesc() const { return mFlagDesc; }
+
+    // Child knobs are displayed in the UI below their parent and slightly indented
+    void               AddChild(Knob* pChild) { mChildren.push_back(pChild); }
+    std::vector<Knob*> GetChildren() const { return mChildren; }
+
+    virtual void Draw() = 0;
+
+    // Flag description in usage message
+    virtual std::string FlagText() const = 0;
+
+    // Updates knob value from commandline flag
+    virtual Result UpdateFromFlag(const CliOptions& opts) = 0;
+
+    // Default values are the values when the application starts. The knobs can be reset to default by a button in the UI
+    virtual void ResetToDefault() = 0;
+
+protected:
+    std::string        mDisplayName;
+    std::string        mFlagName;
+    std::string        mFlagDesc;
+    std::vector<Knob*> mChildren;
+};
+
+// KnobBoolCheckbox is a boolean knob that will be displayed as a checkbox in the UI
+class KnobBoolCheckbox
+    : public Knob
+{
+public:
+    KnobBoolCheckbox(const std::string& displayName, const std::string& flagName, const std::string& flagDesc, bool defaultValue, std::function<void(bool)> callback = NULL)
+        : Knob(displayName, flagName, flagDesc)
+    {
+        mCallback = callback;
+        SetValue(defaultValue, true);
+    }
+
+    void Draw() override;
+    void ResetToDefault() override { SetValue(mDefaultValue); }
+
+    // FlagText format:
+    // --flag_name <true|false> : flag description
+    std::string FlagText() const override;
+
+    // Updates mValue and mDefaultValue, expected flag format:
+    // --flag_name <true|false>
+    Result UpdateFromFlag(const CliOptions& opts) override;
+
+    bool GetValue() const { return mValue; }
+
+    // Used for when mValue needs to be updated outside of UI
+    // If newVal is valid:
+    // - If updateDefault, mDefaultValue is updated
+    // - If newVal is different from mValue:
+    //   - mValue is updated
+    //   - mCallback is triggered
+    void SetValue(bool newVal, bool updateDefault = false);
+
+private:
+    bool mValue;
+    bool mDefaultValue;
+
+    // mCallback will be called on the new bool value every time it is toggled in the UI, or updated with SetValue()
+    std::function<void(bool)> mCallback;
+};
+
+// KnobIntSlider is an int knob that will be displayed as a slider in the UI
+// ImGui sliders can also become input boxes with ctrl + right click
+class KnobIntSlider
+    : public Knob
+{
+public:
+    KnobIntSlider(const std::string& displayName, const std::string& flagName, const std::string& flagDesc, int defaultValue, int minValue, int maxValue, std::function<void(int)> callback = NULL)
```

Comment:
Can we have a SetFlagDesc and SetDisplayName instead of putting it in the constructor?
As far as I know, the only thing that is strictly required is flagName.